design/continuous-delivery.md

```diff
@@ -0,0 +1,332 @@
+```
+
+Higher-level API:
+
+```
+actionName: string;
+assembly: cp.Artifact;     // build output
+frameworkVersion?: string; // defaults to LATEST
+stack: core.Stack;         // the stack to deploy
+```
+
+NOTES:
+
+* The project must run in **privileged** mode in order to support docker
+  assets.
+* The **--exclusively** flag will ensure that “cdk deploy” will only deploy the
+  specified stack (along with it’s assets). This means that if this stack
+  depends on other stacks, they must be explicitly deployed before it.
+* The **--require-approval=never** flag will not require users to confirm any
+  posture changes.
+* The role will be assumed with an external ID dedicated to the CDK to reduce
+  the chance for mistakes.
+
+## Bootstrapping
+
+Based on this model, in order to be able to deploy CDK stacks to an environment
+we need two resources:
+
+1. **Assets bucket**, which is needed by the CLI in order to upload
+   CloudFormation templates and assets to S3.
+2. **Deployment IAM role**, which is needed by the deployment action in order to
+   deploy to this account. The deployment role needs to trust the central
+   account assume-role permissions.
+
+To CDK CLI currently has a `bootstrap` command which provisions the assets
+bucket described above. We will extend it to include a deployment role with an
+option to specify a set of trusted AWS account IDs which will be permitted to
+deploy to this environment:
+
+```
+$ cdk bootstrap aws://account/region --trust-account ACCOUNT_ID
+WARNING: any principal from <ACCOUNT_ID> will have administrative access to the account <account>.
+Please confirm (Y/N):
+```
+
+An appropriate error should be displayed to the user if they try to deploy to a
+non-bootstrapped environment.
+
+## Sample Walkthrough
+
+Given these two primitives (a **stack deployment CodePipeline action** and a
+**bootstrap deployment IAM role**), we are able to support all requirements in
+this doc.
+
+Lets walk through the following scenario:
+
+1. Repository hosted on GitHub.
+2. Separate AWS account for the pipeline.
+3. App includes a VPC stack and a service stack
+4. Initially deploy into a single geography (account/region)
+5. Extend to another geography
+
+### Initialize the pipeline
+
+1. Create an AWS account for the pipeline (we will refer to it as `ACNT-PIPE`)
+2. Bootstrap the pipeline environments:
+
+        cdk bootstrap aws://ACNT-PIPE/us-east-1
+
+3. Create a new git repository for the app (e.g. on GitHub).
+4. Initialize a new CDK app: `cdk init`
+5. Define a **"pipeline"** stack ([example](https://github.com/eladb/cdkcd-test/blob/master/lib/pipeline.ts))
+   which includes a source, build and self-updating pipeline stage.
+6. Commit this code and push to GitHub
+7. Execute `cdk deploy pipeline` to deploy this pipeline.
+8. From this point forward, no need to use `cdk deploy` anymore.
+
+### North America deployment
+
+1. Create an AWS account for the service in North America (`ACNT-NA`)
```

Comment:
It looks like this POC requires the manual creation of a third party role with temp credentials which you are supplying an EXTERNAL_ID for. I love the idea of passing temp credentials to the project build level!

How does everyone feel about having CDK generate this role and an external id?